Fix test memory leak

grcevski opened the pull request:
I had incorrectly made the assumption that the createComponents will only be called once for ILM plugin initialization. It appears that this isn't the case when we run tests.

This PR fixes how we initialize the ILM immutable state handlers, if the LifecyclePlugin gets re created, only the latest state is reflected in the immutable state handler provider.

Fixes https://github.com/elastic/elasticsearch/issues/88337

Commit:
Fix memory leak

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

rjernst commented:
Is there a way we can avoid making production code mutable just for tests?

rjernst commented:
I think the issue is the use of statics. We don't use them in other places for such registration. These handlers should be captured in instance variables specific to the node, so that many node instances (in tests) can be created.

Commit:
Remove static requirement

Commit:
Better naming

grcevski commented:
So I flipped this around a bit and I removed the use of statics. I also added an example code of how this is intended to work in ActionModule.

Essentially, the SPI creates the service provides independently of the plugin constructor, so if we want to avoid any statics, we need a way to pass down what the handlers need to be constructed with. However, the best place to construct the handlers is during the plugin constructor (or createComponents), since then we have all what we'd need to make the handler work.

To avoid passing a superset of what we might need for construction of the immutable state handlers, I did the following:

- Let the plugin create the immutable handler just like any other action handler during construction and keep a reference to it
- When and if a handler provider is created, it will ask the plugin to deliver a collection of the handlers it likes to expose.
- The provider now receives a list of plugins it 'serves' and it asks the plugins for what handlers they'll expose. We have the plugin instance at the time we call for the providers to be created, since we have already loaded the plugins by then.

grcevski commented on server/src/main/java/org/elasticsearch/immutablestate/ImmutableClusterStateHandlerProvider.java:
```diff
@@ -21,5 +23,7 @@ public interface ImmutableClusterStateHandlerProvider {
-    Collection<ImmutableClusterStateHandler<?>> handlers();
+    Collection<ImmutableClusterStateHandler<?>> handlers(Collection<? extends Plugin> loadedPlugins);
+
+    Collection<Class<? extends Plugin>> providedPlugins();
```
I need to add docs here.

grcevski commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/ILMImmutableStateHandlerProvider.java:
```diff
@@ -9,24 +9,30 @@
     @Override
-    public Collection<ImmutableClusterStateHandler<?>> handlers() {
-        return handlers;
+    public Collection<ImmutableClusterStateHandler<?>> handlers(Collection<? extends Plugin> loadedPlugins) {
```
The provider receives a reference back of its related plugins, which will supply it with what handlers they want to expose.

grcevski commented:
I think also, maybe I don't need to worry about a collection of plugins here. It should be probably be only one per plugin.

Commit:
Simplify with 1:1 mapping

Commit:
Use our createExtensions SPI

grcevski commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/ILMImmutableStateHandlerProvider.java:
```diff
@@ -10,23 +10,24 @@
-    @Override
-    public Collection<ImmutableClusterStateHandler<?>> handlers() {
-        return handlers;
+    public ILMImmutableStateHandlerProvider() {
```
Declaring this constructor to make sure module-info compiles properly, since SPI requires that we use a no-arg constructor. If we switch to using the META-INF/services file, this problem goes away, however our build 'validateModule' task fails, requiring that we also expose this service provider in our module provides section.

rjernst commented on server/src/main/java/org/elasticsearch/action/ActionModule.java:
```diff
@@ -888,6 +892,23 @@ public void initRestHandlers(Supplier<DiscoveryNodes> nodesInCluster) {
+
+        handlers.add(new ImmutableClusterSettingsAction(clusterSettings));
+        pluginHandlers.forEach(h -> handlers.addAll(h.handlers()));
+        // Initialize the controller when merged
```
The two collections here seem to be locals. Shouldn't they be captured somewhere?

grcevski replied:
Yeah, this is unused, I brought in the code to demonstrate how we'd use these when I had the earlier approach. I'll remove this code altogether. The values are locals because the initialization of the controller is not here yet, as in the other PR: https://github.com/elastic/elasticsearch/pull/88224/files#diff-b56bed42f5b0025886eb243ecb48678ac048bfba0ef047545dbb6504e357edf2R915.

I agree with your comments, so I'll refactor the code in the other PR to apply your suggestions on how to initialize the immutable handlers in the other PR.

rjernst commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -1055,6 +1055,8 @@ protected Node(
             logger.debug("initializing HTTP handlers ...");
             actionModule.initRestHandlers(() -> clusterService.state().nodesIfRecovered());
+            logger.debug("initializing operator handlers ...");
+            actionModule.initImmutableClusterStateHandlers(pluginsService);
```
This doesn't seem like the right place in node construction to init these. The action module is here in the line above for other reasons (and can actually be moved up, it only needs the cluster service created). I would init these when the action module is created, as part of the ctor.

rjernst commented on server/src/main/java/org/elasticsearch/action/ActionModule.java:
```diff
@@ -888,6 +892,23 @@ public void initRestHandlers(Supplier<DiscoveryNodes> nodesInCluster) {
+    public void initImmutableClusterStateHandlers(PluginsService pluginsService) {
+        List<ImmutableClusterStateHandler<?>> handlers = new ArrayList<>();
+
+        List<? extends ImmutableClusterStateHandlerProvider> pluginHandlers = pluginsService.loadServiceProviders(
```
I think I would rather see loadServiceProviders called in Node, and pass in the List of providers to ActionModule. That better matches the other calls in Node to eg filterPlugins. In general, we try to avoid passing the plugin service itself.

rjernst commented on server/src/test/java/org/elasticsearch/plugins/PluginsServiceTests.java:
```diff
@@ -531,6 +531,10 @@ public TestExtension() {}
 
             }
+
+            public TestExtension(TestPlugin plugin, String anotherArg) {
```
Is this used somewhere?

grcevski replied:
Yes, I added a constructor to make the test check that we disallow a service provider with more than 2 constructors. Previously this class was testing for existence of more than 1.

rjernst reviewed (approved):
I left a couple asks on how this looks with Node/ActionModule. LGTM if you agree on the points I left.

Commit:
Remove unused code

grcevski commented:
@elasticmachine run elasticsearch-ci/part-2
